src/Caching/SqlServer/src/SqlServerCacheOptions.cs

```diff
@@ -43,6 +43,11 @@ public class SqlServerCacheOptions : IOptions<SqlServerCacheOptions>
         /// </summary>
         public TimeSpan DefaultSlidingExpiration { get; set; } = TimeSpan.FromMinutes(20);
 
+        /// <summary>
+        /// A function, if provided, should return an access token for authentication
+        /// </summary>
+        public Func<string> AccessTokenFunc { get; set; }
```

Comment:
```suggestion
        public Func<string> AccessTokenProvider { get; set; }
```

This name aligns with what we use in other places. Also, I think it's **good** that this isn't async. We may need to remind users that they will have to fetch the access token asynchronously in a background component and swap the new token into a singleton value of some kind since they can't dynamically fetch an access token in the handler (because it's synchronous).

Reply:
Applied the name change as you suggested. This is the implementation that I had in mind for the AccessTokenProvider:  
```
AccessTokenProvider = () => {
  var azureServiceTokenProvider = new AzureServiceTokenProvider();

  return azureServiceTokenProvider.GetAccessTokenAsync("https://database.windows.net/").GetAwaiter().GetResult();
}
```

Reply:
I was afraid you'd say that :). Calling `.GetResult()` on an awaiter or `.Wait`/`.Result` on a task will tank performance and make your app very difficult to scale. Blocking waits on Tasks is always a **bad idea**.

The scalable way to do this would be to have a background service call [`GetAuthenticationResultAsync`](https://github.com/Azure/azure-sdk-for-net/blob/0bf39aa206dbbe867619f4a4458d698d9e058e83/src/SdkCommon/AppAuthentication/Azure.Services.AppAuthentication/AzureServiceTokenProvider.cs#L264), store the token somewhere the callback can access and set a timer that will fire when the token is near expiry. Then when the timer fires, get a new token and atomically update the stored access token so that the next SQL connection uses it.

I think we're going to need to look in to this further to find the right design to make sure the SQL cache can use Azure Managed Service Identity properly.

Reply:
hahaha :) this is the solution I have seen floating around for enabling Managed Identity for use with Entity Framework in the community. Internally the `AzureServiceTokenProvider` caches the token and refreshes if it has expired or is about to expire (5 minutes) so I figured I could reply on it since they're already taking care of multi-thread access. But I see your point about the potential performance hit if implemented as I suggested.

Reply:
This is the implementation of the `AzureServiceTokenProvider` [Source Code](https://github.com/Azure/azure-sdk-for-net/blob/master/src/SdkCommon/AppAuthentication/Azure.Services.AppAuthentication/AzureServiceTokenProvider.cs)

Let me know your thoughts if the `Awaiter` in this case will be okay or just simply not a good idea :)

Reply:
It's still universally a Bad Idea:tm: to call `.GetAwaiter().GetResult()`. Even if the underlying awaiter is "rarely" async. It's just a matter of good practice.

Reply:
> this is the solution I have seen floating around for enabling Managed Identity for use with Entity Framework in the community

This is unfortunate. cc @ajcvickers for visibility. Might be good for the EF team to try to get out some "official" docs or guidance for using EF with [Azure Managed Service Identity](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/tutorial-windows-vm-access-sql) if there aren't already.